azure-spring-boot/src/main/java/com/microsoft/azure/spring/autoconfigure/b2c/UserPrincipal.java

```diff
@@ -0,0 +1,189 @@
+
+    private static final String ATTRIBUTE_POSTAL_CODE = "postalCode";
+
+    private static final String ATTRIBUTE_JOB_TITLE = "jobTitle";
+
+    private static final String ATTRIBUTE_GIVEN_NAME = "given_name";
+
+    private static final String ATTRIBUTE_DISPLAY_NAME = "name";
+
+    private static final String ATTRIBUTE_COUNTRY = "country";
+
+    private static final String ATTRIBUTE_CITY = "city";
+
+    private static final String ATTRIBUTE_NEW_USER = "newUser";
+
+    private static final String ATTRIBUTE_TRUST_FRAMEWORK_POLICY = "tfp";
+
+    private static final String ATTRIBUTE_EMAILS = "emails";
+
+    private static final String ATTRIBUTE_CODE_HASH = "c_hash";
+
+    private static final String ATTRIBUTE_NONCE = "nonce";
+
+    private static final String ATTRIBUTE_ID_PROVIDER = "idp";
+
+    // JWT token
+    private final String keyId;
+
+    private final String algorithm;
+
+    private final String subject;
+
+    private final String issuer;
+
+    private final Date expireDate;
+
+    private final Date notValidBeforeDate;
+
+    private final Date issueAtDate;
+
+    private final String nonce;
+
+    private final List<String> audiences;
+
+    private final Map<String, Object> claim;
+
+    private final String codeHash;
+
+    private final String code; // The authorization code.
+
+    // AADB2C pre-defined User attributes.
+    private final String city;
+
+    private final String countryOrRegion;
+
+    private final String displayName;
+
+    private final List<String> emails;
+
+    private final String givenName;
+
+    private final String identityProvider;
+
+    private final String jobTitle;
+
+    private final String legalAgeGroup;
+
+    private final String postalCode;
+
+    private final String stateOrProvince;
+
+    private final String street;
+
+    private final String surname;
+
+    private final Boolean isNew;
+
+    private final String userObjectId;
+
+    private final String trustFrameworkPolicy;
+
+    @SuppressWarnings("unchecked")
+    public UserPrincipal(@NonNull Pair<JWSObject, JWTClaimsSet> jwtToken, String code) {
+        final JWSObject jwsObject = jwtToken.getLeft();
+        final JWTClaimsSet jwtClaimsSet = jwtToken.getRight();
+
+        this.keyId = jwsObject.getHeader().getKeyID();
+        this.algorithm = jwsObject.getHeader().getAlgorithm().getName();
+        this.subject = jwtClaimsSet.getSubject();
+        this.issuer = jwtClaimsSet.getIssuer();
+        this.audiences = jwtClaimsSet.getAudience();
+        this.expireDate = jwtClaimsSet.getExpirationTime();
+        this.notValidBeforeDate = jwtClaimsSet.getNotBeforeTime();
+        this.issueAtDate = jwtClaimsSet.getIssueTime();
+        this.claim = jwtClaimsSet.getClaims();
+        this.code = code;
+
+        this.userObjectId = jwtClaimsSet.getSubject(); // Subject populates the object ID of the user in the directory.
+
+        this.surname = (String) getClaim().get(ATTRIBUTE_SURNAME);
```

Comment:
Cache the result of `getClaim()`?

Reply:
seems unnecessary here, as just a Getter.